internal/controllers/topology/cluster/cluster_controller.go

```diff
@@ -173,9 +175,26 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		return ctrl.Result{}, err
 	}
 
+	// Get ClusterClass.
+	clusterClass := &clusterv1.ClusterClass{}
+	key := client.ObjectKey{Name: cluster.Spec.Topology.Class, Namespace: cluster.Namespace}
+	if err := r.Client.Get(ctx, key, clusterClass); err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to retrieve ClusterClass %s", cluster.Spec.Topology.Class)
+	}
+	// Default and Validate the Cluster based on information from the ClusterClass.
+	// This step is needed as if the ClusterClass does not exist at Cluster creation some fields may not be defaulted or
+	// validated in the webhook.
+	if errs := webhooks.DefaultVariables(cluster, clusterClass); len(errs) > 0 {
+		return ctrl.Result{}, apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("Cluster").GroupKind(), cluster.Name, errs)
+	}
```

Comment:
Hm maybe it's fine. Did a few tests:
* regular webhook defaulting seems to give ownership to the client (aka kubectl)
* when a user deploys a Cluster CR, the user owns the entire variables field
* when the controller does defaulting it takes over ownership of the entire field
* if a user re-applies a YAML they get the ownership for the entire field again (and they can drop fields that way)

So my initial assessment would be that the current behavior is probably okay as the variables slice is atomically merged. Because of that adding/removing/updating variables just works as expected. I think we would run into more problems if we would have fine-granular ownership tracking.

Reply:
I think in contrast to the previous case we had with SSA (AWSCluster) we don't have a strict co-ownership requirement.

I would assume it could happen that the user (via GitOps)  and the cluster controller are fighting over the write value of the variables slice. But this can be easily resolved by the user including the defaulted values in the YAML they apply.

If that's not good enough we would have to explore if we can change the variables slice to a list of type map. The problem there is the unique map key. I don't know if we can use namespace+name as a map key, given that namespace can be empty.

I would also assume that if we use a list map we run into problems when our controller takes over ownership of individual problems (i.e. the user is then unable to remove variables that are owned by the controller).

Reply:
> I would assume it could happen that the user (via GitOps) and the cluster controller are fighting over the right value of the variables slice. But this can be easily resolved by the user including the defaulted values in the YAML they apply.

Discussed with Killian. This should be actually fine as sooner or later the defaulting webhook will add the default values and then the user will own them

Reply:
Patch will give ownership to "manager". Except you mean we want to clear the managed fields with a second patch call

Reply:
Yeah - I'm not sure I understand if there's a way we can avoid the SSA patch here. I have a follow up to come back to the SSA story, and to test it properly, once we have the basics in place.

Reply:
Yup let's follow-up on that and then think through our options.